File: chat-uikit/src/main/java/io/agora/chat/uikit/adapter/EaseBaseChatExtendMenuAdapter.java
package io.agora.chat.uikit.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public abstract class EaseBaseChatExtendMenuAdapter<VH extends RecyclerView.ViewHolder, T> extends RecyclerView.Adapter<VH> {
    public List<T> mData;

    public void setData(List<T> data) {
        mData = data;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mData == null ? 0 : mData.size();
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(getItemLayoutId(), parent, false);
        return easeCreateViewHolder(view);
    }

    protected abstract int getItemLayoutId();

    /**
     * Get ViewHolder
     * @param view
     * @return
     */
    protected abstract VH easeCreateViewHolder(View view);

}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/base/EaseBaseChatExtendMenuAdapter.kt
package io.agora.uikit.base

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

abstract class EaseBaseChatExtendMenuAdapter<VH : RecyclerView.ViewHolder?, T> :
    RecyclerView.Adapter<VH>() {
    protected var mData: List<T>? = null
    fun setData(data: List<T>?) {
        mData = data
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return if (mData == null) 0 else mData!!.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH & Any {
        val view = LayoutInflater.from(parent.context).inflate(itemLayoutId, parent, false)
        return easeCreateViewHolder(view)
    }

    protected abstract val itemLayoutId: Int

    /**
     * Get ViewHolder
     * @param view
     * @return
     */
    protected abstract fun easeCreateViewHolder(view: View?): VH & Any
}